### PR TITLE
OFI-NCCL [Experimental]: Add P4De topology

### PR DESCRIPTION
**Pull request by rashikakheria**
This patch adds the right system topology for P4De platform. The
topology file is very similar to already available P4D topology
with minor changes to GPUs device IDs (80G support).

This patch also introduces refactoring to topology update logic
to enable less painful addition of new platforms.

Signed-off-by: Rashika Kheria <rashika@amazon.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/nccl_ofi_net.c [hunk 1]
```diff
@@ -1050,49 +1068,120 @@ static inline ncclResult_t nccl_ofi_progress(nccl_ofi_t *nccl_ofi_comp)
...
 
 	fd = fopen(file, "r");
 	if (fd == NULL) {
 		NCCL_OFI_WARN("Error opening file: %s", file);
 		return -1;
 	}
 
-	/* Read first line of the file */
+	/* Read first line of the file, reallocing the buffer as necessary */
 	while ((feof(fd) == 0) && (ferror(fd) == 0) && ((ch = fgetc(fd)) != '\n')) {
```
Comment (nzmsv) [inline src/nccl_ofi_net.c:1095]: Could this not just be an `fscanf` with a max length?

File: src/nccl_ofi_net.c [hunk 2]
```diff
@@ -1050,49 +1068,120 @@ static inline ncclResult_t nccl_ofi_progress(nccl_ofi_t *nccl_ofi_comp)
...
+ * @brief	Returns static topology filename for given platform type, if found
+ *
+ * @input	Platform type
+ *
+ * @return	NULL, if no topology found
+ * 		Topology filename, if match found
+ */
+static char* get_static_topology_file(char *platform_type)
+{
+	int platform_n = sizeof(platform_topo_map)/sizeof(platform_topo_map[0]);
```
Comment (nzmsv) [inline src/nccl_ofi_net.c:1119]: `static const int` so it's compile-time?

File: src/nccl_ofi_net.c [hunk 3]
```diff
@@ -1050,49 +1068,120 @@ static inline ncclResult_t nccl_ofi_progress(nccl_ofi_t *nccl_ofi_comp)
...
+	} else {
+		/* Update topology */
+		char topology_path[PATH_MAX];
 
-		/* Report mismatch if characters read are more than string length */
-		if (idx == len) {
-			ret = 0;
+		rc = snprintf(topology_path, sizeof(topology_path), "%s/%s",
+				XML_DIR, topo_file);
+		if (rc < 0 || rc >= PATH_MAX) {
```
Comment (nzmsv) [inline src/nccl_ofi_net.c:1163]: Nit: `rc >= sizeof(topology_path)`

File: src/nccl_ofi_net.c (hunk 1 above)
Comment (nzmsv) [reply src/nccl_ofi_net.c:1095]: I'm OK with shipping this, but I'll provide the other side of the argument.

I think it's extremely unlikely that a machine model is longer than 128 bytes.  If it happens to be, there's likely another bug somewhere and it would be better to simply error out the file read rather than allocating potentially unbounded memory.  

And as written the code actually encourages a much more likely programming error: it requires a buffer that has been `malloc`ed to be passed in.  Someone could easily pass in a stack or global buffer.  Usually the functions that return strings in this style do the `malloc` internally and then require the user to free the buffer.  Here we are mixing things up: `malloc` is on the outside, but `realloc` is on the inside.

Basically we traded a very unlikely footgun for one that is very likely to fire :)
